// @aws-solutions-constructs/core/test/dynamo-table.test.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const aws_cdk_lib_1 = require("aws-cdk-lib");
const dynamodb = require("aws-cdk-lib/aws-dynamodb");
const defaults = require("../index");
const utils_1 = require("../lib/utils");
const assertions_1 = require("aws-cdk-lib/assertions");
const dynamodb_table_helper_1 = require("../lib/dynamodb-table-helper");
test('test TableProps change billing mode', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const defaultProps = defaults.GetDefaultTableProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        },
    });
    const inProps = {
        billingMode: dynamodb.BillingMode.PROVISIONED,
        readCapacity: 3,
        writeCapacity: 3,
        partitionKey: {
            name: 'id',
            type: dynamodb.AttributeType.STRING
        }
    };
    const outProps = (0, utils_1.overrideProps)(defaultProps, inProps);
    new dynamodb.Table(stack, 'test-dynamo-override', outProps);
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties("AWS::DynamoDB::Table", {
        KeySchema: [
            {
                AttributeName: "id",
                KeyType: "HASH"
            }
        ],
        AttributeDefinitions: [
            {
                AttributeName: "id",
                AttributeType: "S"
            }
        ],
        ProvisionedThroughput: {
            ReadCapacityUnits: 3,
            WriteCapacityUnits: 3
        },
        SSESpecification: {
            SSEEnabled: true
        }
    });
});
test('test TableProps override add sort key', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const defaultProps = defaults.GetDefaultTableProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        },
    });
    const inProps = {
        partitionKey: {
            name: 'id',
            type: dynamodb.AttributeType.STRING
        },
        sortKey: {
            name: 'sort_key',
            type: dynamodb.AttributeType.STRING
        }
    };
    const outProps = (0, utils_1.overrideProps)(defaultProps, inProps);
    new dynamodb.Table(stack, 'test-dynamo-override', outProps);
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties("AWS::DynamoDB::Table", {
        KeySchema: [
            {
                AttributeName: "id",
                KeyType: "HASH"
            },
            {
                AttributeName: "sort_key",
                KeyType: "RANGE"
            }
        ],
        AttributeDefinitions: [
            {
                AttributeName: "id",
                AttributeType: "S"
            },
            {
                AttributeName: "sort_key",
                AttributeType: "S"
            }
        ],
        BillingMode: "PAY_PER_REQUEST",
        SSESpecification: {
            SSEEnabled: true
        }
    });
});
test('test TableWithStreamProps override stream view type', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const defaultProps = defaults.GetDefaultTableWithStreamProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        },
    });
    const inProps = {
        partitionKey: {
            name: 'id',
            type: dynamodb.AttributeType.STRING
        },
        stream: dynamodb.StreamViewType.NEW_IMAGE
    };
    const outProps = (0, utils_1.overrideProps)(defaultProps, inProps);
    new dynamodb.Table(stack, 'test-dynamo-override', outProps);
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties("AWS::DynamoDB::Table", {
        KeySchema: [
            {
                AttributeName: "id",
                KeyType: "HASH"
            }
        ],
        AttributeDefinitions: [
            {
                AttributeName: "id",
                AttributeType: "S"
            }
        ],
        BillingMode: "PAY_PER_REQUEST",
        SSESpecification: {
            SSEEnabled: true
        },
        StreamSpecification: {
            StreamViewType: "NEW_IMAGE"
        }
    });
});
test('test buildDynamoDBTable with existingTableObj', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const tableProps = {
        billingMode: dynamodb.BillingMode.PROVISIONED,
        partitionKey: {
            name: 'table_id',
            type: dynamodb.AttributeType.STRING
        }
    };
    const existingTableObj = new dynamodb.Table(stack, 'DynamoTable', tableProps);
    const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {
        existingTableObj
    });
    expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
    expect(buildDynamoDBTableResponse.tableObject).toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "table_id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5
        }
    });
});
test('test buildDynamoDBTable without any arguments', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {});
    expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
    expect(buildDynamoDBTableResponse.tableObject).toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        BillingMode: "PAY_PER_REQUEST"
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        SSESpecification: {
            SSEEnabled: true
        }
    });
});
test('test buildDynamoDBTable with TableProps', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const dynamoTableProps = {
        billingMode: dynamodb.BillingMode.PROVISIONED,
        partitionKey: {
            name: 'table_id',
            type: dynamodb.AttributeType.STRING
        }
    };
    const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {
        dynamoTableProps
    });
    expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
    expect(buildDynamoDBTableResponse.tableObject).toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "table_id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5
        }
    });
});
test('test buildDynamoDBTableWithStream with TableProps', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const dynamoTableProps = {
        partitionKey: {
            name: 'table_id',
            type: dynamodb.AttributeType.STRING
        },
        stream: dynamodb.StreamViewType.NEW_IMAGE
    };
    const response = defaults.buildDynamoDBTableWithStream(stack, {
        dynamoTableProps
    });
    expect(response.tableInterface).toBeDefined();
    expect(response.tableObject).toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "table_id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        StreamSpecification: {
            StreamViewType: "NEW_IMAGE"
        }
    });
});
test('test buildDynamoDBTableWithStream without any arguments', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const response = defaults.buildDynamoDBTableWithStream(stack, {});
    expect(response.tableInterface).toBeDefined();
    expect(response.tableObject).toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        BillingMode: "PAY_PER_REQUEST"
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        SSESpecification: {
            SSEEnabled: true
        }
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        StreamSpecification: {
            StreamViewType: "NEW_AND_OLD_IMAGES"
        }
    });
});
test('test buildDynamoDBTableWithStream with existingTableObj', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const tableProps = {
        partitionKey: {
            name: 'table_id',
            type: dynamodb.AttributeType.STRING
        },
        stream: dynamodb.StreamViewType.NEW_IMAGE
    };
    const existingTableInterface = new dynamodb.Table(stack, 'DynamoTable', tableProps);
    const response = defaults.buildDynamoDBTableWithStream(stack, {
        existingTableInterface
    });
    expect(response.tableInterface).toBeDefined();
    expect(response.tableObject).not.toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "table_id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        StreamSpecification: {
            StreamViewType: "NEW_IMAGE"
        }
    });
});
test('test buildDynamoDBTable with existingTableInterface', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const tableProps = {
        partitionKey: {
            name: 'table_id',
            type: dynamodb.AttributeType.STRING
        },
        stream: dynamodb.StreamViewType.NEW_IMAGE
    };
    const existingTableInterface = new dynamodb.Table(stack, 'DynamoTable', tableProps);
    const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {
        existingTableInterface
    });
    expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
    expect(buildDynamoDBTableResponse.tableObject).not.toBeDefined();
    const template = assertions_1.Template.fromStack(stack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        KeySchema: [
            {
                AttributeName: "table_id",
                KeyType: "HASH"
            }
        ]
    });
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        StreamSpecification: {
            StreamViewType: "NEW_IMAGE"
        }
    });
});
test('test getPartitionKeyNameFromTable()', () => {
    const partitionKeyName = 'testPartitionKey';
    const stack = new aws_cdk_lib_1.Stack();
    const defaultProps = defaults.GetDefaultTableProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        },
    });
    const inProps = {
        partitionKey: {
            name: partitionKeyName,
            type: dynamodb.AttributeType.STRING
        },
        sortKey: {
            name: 'sort_key',
            type: dynamodb.AttributeType.STRING
        }
    };
    const outProps = (0, utils_1.overrideProps)(defaultProps, inProps);
    const newTable = new dynamodb.Table(stack, 'test-dynamo-override', outProps);
    const testKeyName = (0, dynamodb_table_helper_1.getPartitionKeyNameFromTable)(newTable);
    expect(testKeyName).toEqual(partitionKeyName);
});
test('Test that PointInTimeRecovery is not set when PointInTimeRecoverySpecification is provided', () => {
    const defaultProps = defaults.GetDefaultTableProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        },
        pointInTimeRecoverySpecification: {
            pointInTimeRecoveryEnabled: true
        }
    });
    expect(defaultProps.pointInTimeRecovery).toBeUndefined();
    // Let's confirm the alternative  as well
    const moreDefaultProps = defaults.GetDefaultTableProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        }
    });
    expect(moreDefaultProps.pointInTimeRecovery).toEqual(true);
});
test('Test that PointInTimeRecovery is not set when PointInTimeRecoverySpecification is provided for table with stream', () => {
    const defaultProps = defaults.GetDefaultTableWithStreamProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        },
        pointInTimeRecoverySpecification: {
            pointInTimeRecoveryEnabled: true
        }
    });
    expect(defaultProps.pointInTimeRecovery).toBeUndefined();
    // Let's confirm the alternative  as well
    const moreDefaultProps = defaults.GetDefaultTableWithStreamProps({
        partitionKey: {
            name: 'donotuse',
            type: dynamodb.AttributeType.BINARY
        }
    });
    expect(moreDefaultProps.pointInTimeRecovery).toEqual(true);
});
// ---------------------------
// Prop Tests
// ---------------------------
test('Test fail DynamoDB table check', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const props = {
        existingTableObj: new dynamodb.Table(stack, 'placeholder', defaults.GetDefaultTableProps({
            partitionKey: {
                name: 'donotuse',
                type: dynamodb.AttributeType.BINARY
            },
        })),
        dynamoTableProps: defaults.GetDefaultTableProps({
            partitionKey: {
                name: 'donotuse',
                type: dynamodb.AttributeType.BINARY
            },
        }),
    };
    const app = () => {
        defaults.CheckDynamoDBProps(props);
    };
    // Assertion
    expect(app).toThrowError('Error - Either provide existingTableObj or dynamoTableProps, but not both.\n');
});
test('Test fail DynamoDB table check (for interface AND obj)', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const props = {
        existingTableInterface: new dynamodb.Table(stack, 'placeholder', defaults.GetDefaultTableProps({
            partitionKey: {
                name: 'donotuse',
                type: dynamodb.AttributeType.BINARY
            },
        })),
        existingTableObj: new dynamodb.Table(stack, 'placeholderobj', defaults.GetDefaultTableProps({
            partitionKey: {
                name: 'donotuse',
                type: dynamodb.AttributeType.BINARY
            },
        })),
    };
    const app = () => {
        defaults.CheckDynamoDBProps(props);
    };
    // Assertion
    expect(app).toThrowError('Error - Either provide existingTableInterface or existingTableObj, but not both.\n');
});
