// @aws-solutions-constructs/core/lib/kinesis-firehose-s3-defaults.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultCfnDeliveryStreamProps = DefaultCfnDeliveryStreamProps;
/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
function DefaultCfnDeliveryStreamProps(bucketArn, roleArn, logGroupName, logStreamName, kms, deliveryStreamName) {
    return {
        deliveryStreamName,
        extendedS3DestinationConfiguration: {
            bucketArn,
            bufferingHints: {
                intervalInSeconds: 300,
                sizeInMBs: 5
            },
            compressionFormat: 'GZIP',
            roleArn,
            cloudWatchLoggingOptions: {
                enabled: true,
                logGroupName,
                logStreamName
            },
            encryptionConfiguration: {
                kmsEncryptionConfig: {
                    awskmsKeyArn: kms.keyArn
                }
            }
        },
    };
}
