// @aws-solutions-constructs/core/lib/pipes-defaults.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultPipesProps = defaultPipesProps;
exports.defaultSqsSourceProps = defaultSqsSourceProps;
exports.defaultDynamoDBStreamsSourceProps = defaultDynamoDBStreamsSourceProps;
exports.defaultStateMachineTargetProps = defaultStateMachineTargetProps;
// import * as pipes from 'aws-cdk-lib/aws-pipes';
function defaultPipesProps() {
    return {};
}
function defaultSqsSourceProps() {
    return {};
}
function defaultDynamoDBStreamsSourceProps(deployDlq) {
    return {
        dynamoDbStreamParameters: {
            startingPosition: "LATEST",
            maximumRetryAttempts: deployDlq ? 3 : undefined,
        }
    };
}
function defaultStateMachineTargetProps() {
    return {
        stepFunctionStateMachineParameters: {
            invocationType: 'FIRE_AND_FORGET'
        }
    };
}
