// @aws-solutions-constructs/core/lib/s3-oac-origin.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3OacOrigin = void 0;
/*
 *  The functions found here in the core library are for internal use and can be changed
 *  or removed outside of a major release. We recommend against calling them directly from client code.
 */
const cloudfront = require("aws-cdk-lib/aws-cloudfront");
const aws_cloudfront_origins_1 = require("aws-cdk-lib/aws-cloudfront-origins");
const utils_1 = require("./utils");
/**
 * A custom implementation of S3Origin that allows an origin access control (OAC) to be used instead of
 * an origin access identity (OAI), which is currently the only option supported by default CDK.
 */
class S3OacOrigin {
    constructor(bucket, props) {
        if (bucket.isWebsite) {
            // If the bucket is configured for website hosting, set up an HttpOrigin to support legacy clients
            (0, utils_1.printWarning)(`Bucket ${bucket.bucketName} is being provided as a source but currently has website hosting enabled.
        This requires both the bucket and its objects to be public. AWS strongly recommends against configuring buckets
        and objects for public access. As an alternative, we recommend turning off website hosting settings on the bucket,
        which will result in an origin access control (OAC) being provisioned through which CloudFront can securely
        serve assets from the bucket.`);
            this.origin = new aws_cloudfront_origins_1.HttpOrigin(bucket.bucketWebsiteDomainName, {
                protocolPolicy: cloudfront.OriginProtocolPolicy.HTTP_ONLY, // S3 only supports HTTP for website buckets
                ...props
            });
        }
        else {
            if (!props.originAccessControl) {
                throw new Error(`"props.originAccessControl" is undefined. An origin access control must be provided when using a
          bucket that does not have website hosting enabled.`);
            }
            // If else, set up the origin access control
            this.origin = new S3OacBucketOrigin(bucket, props.originAccessControl);
        }
    }
    bind(scope, options) {
        return this.origin.bind(scope, options);
    }
}
exports.S3OacOrigin = S3OacOrigin;
/**
 * An origin specific to a S3 bucket (not configured for website hosting).
 */
class S3OacBucketOrigin extends cloudfront.OriginBase {
    constructor(bucket, originAccessControl) {
        super(bucket.bucketRegionalDomainName);
        this.originAccessControl = originAccessControl;
    }
    bind(scope, options) {
        return super.bind(scope, options);
    }
    renderS3OriginConfig() {
        return { originAccessIdentity: '' };
    }
}
